File: scripts/upload_to_drive.py
import os
import datetime
import sys
import glob
from googleapiclient.discovery import build
from googleapiclient.http import MediaFileUpload
from google.oauth2 import service_account
from google.oauth2.credentials import Credentials
from google_auth_oauthlib.flow import InstalledAppFlow
from google.auth.transport.requests import Request
from dotenv import load_dotenv
import json

# Load environment variables
load_dotenv(".env.local")

# Wrapper to handle both .env and .env.local
if not os.getenv("GOOGLE_DRIVE_FOLDER_ID"):
    load_dotenv()

PARENT_FOLDER_ID = os.getenv("GOOGLE_DRIVE_FOLDER_ID")
SCOPES = ['https://www.googleapis.com/auth/drive']

def authenticate():
    """Authenticates using User OAuth Token (Cloud/Local) or Service Account."""
    creds = None
    
    # 1. Try User OAuth Token from Environment Variable (GitHub Actions / Cloud)
    token_info = os.getenv("GOOGLE_DRIVE_TOKEN_JSON")
    if token_info:
        print("🔐 Authenticating via User OAuth Token (Env Var)...")
        info = json.loads(token_info)
        creds = Credentials.from_authorized_user_info(info, SCOPES)
        return build('drive', 'v3', credentials=creds)

    # 2. Try Local token.json
    if os.path.exists('token.json'):
        print("👤 Authenticating via Local token.json...")
        creds = Credentials.from_authorized_user_file('token.json', SCOPES)
    
    # Refresh if expired
    if creds and creds.expired and creds.refresh_token:
        print("🔄 Refreshing expired token...")
        try:
            creds.refresh(Request())
            with open('token.json', 'w') as token:
                token.write(creds.to_json())
        except Exception as e:
            print(f"⚠️ Failed to refresh token: {e}")
            creds = None

    if creds and creds.valid:
        return build('drive', 'v3', credentials=creds)

    # 3. Try Service Account (Fallback for specific tasks if needed)
    service_account_info = os.getenv("GOOGLE_SERVICE_ACCOUNT_INFO")
    service_account_path = 'service_account_key.json'
    
    if service_account_info:
        print("🤖 Authenticating via Service Account (Env Var)...")
        info = json.loads(service_account_info)
        creds = service_account.Credentials.from_service_account_info(info, scopes=SCOPES)
    elif os.path.exists(service_account_path):
        print("🤖 Authenticating via Service Account (Key File)...")
        creds = service_account.Credentials.from_service_account_file(service_account_path, scopes=SCOPES)
    
    if creds:
        return build('drive', 'v3', credentials=creds)

    # 4. Final Fallback: Full Interactive Login (Local Only)
    print("❌ No valid credentials found. Starting interactive login...")
    if not os.path.exists('credentials.json'):
        print("❌ Error: credentials.json not found.")
        sys.exit(1)
        
    flow = InstalledAppFlow.from_client_secrets_file('credentials.json', SCOPES)
    creds = flow.run_local_server(port=0)
    with open('token.json', 'w') as token:
        token.write(creds.to_json())

    return build('drive', 'v3', credentials=creds)

def create_folder(service, name, parent_id):
    """Create a folder and return its ID."""
    print(f"Checking/Creating folder: {name}")
    
    # Check if exists first
    query = f"name = '{name}' and '{parent_id}' in parents and mimeType = 'application/vnd.google-apps.folder' and trashed = false"
    # supportsAllDrives=True handles both Shared Drives and standard shared folders better
    results = service.files().list(q=query, fields="files(id, name)", supportsAllDrives=True, includeItemsFromAllDrives=True).execute()
    files = results.get('files', [])
    
    if files:
        print(f"Folder '{name}' already exists. ID: {files[0]['id']}")
        return files[0]['id']
    else:
        file_metadata = {
            'name': name,
            'mimeType': 'application/vnd.google-apps.folder',
            'parents': [parent_id]
        }
        file = service.files().create(body=file_metadata, fields='id', supportsAllDrives=True).execute()
        print(f"Created folder '{name}'. ID: {file.get('id')}")
        return file.get('id')

def upload_file(service, file_path, folder_id):
    name = os.path.basename(file_path)
    
    # Rename Logic for Clarity
    # [slug].mp4 -> 【動画】[slug].mp4
    # [slug].md -> 【台本】[slug] (Doc)
    # [slug]_prompts.md -> 【プロンプト】[slug] (Doc)
    # [slug]_posts.md -> 【SNS】[slug] (Doc)
    
    
    display_name = name
    
    # Path-based detection for accuracy
    if "content/social" in file_path and name.endswith(".md"):
         display_name = f"【SNS】{name}"
    elif "content/prompts" in file_path and name.endswith(".md"):
         display_name = f"【プロンプト】{name}"
    elif "content/scripts" in file_path and name.endswith(".md"):
         display_name = f"【台本】{name}"
    elif "content/articles" in file_path and name.endswith(".md"):
         display_name = f"【記事原稿】{name}" # Article draft
    elif name == "walkthrough.md":
         display_name = "【報告書】walkthrough"
    elif name.endswith(".md"):
        # Fallback for other MD files
        display_name = f"【資料】{name}"
    elif name.endswith("_seed.png"):
        display_name = f"【種画像】{name}"
    elif name.endswith(".mp4"):
        display_name = f"【動画】{name}"

    file_metadata = {
        'name': display_name,
        'parents': [folder_id]
    }
    
    # Auto-convert Markdown to Google Doc for mobile editing
    target_mime = None
    source_mime = None
    
    if file_path.endswith('.md') or (file_path.endswith('.txt') and "テロップ" in name):
        print(f"🔄 Converting to Google Doc: {display_name}")
        target_mime = 'application/vnd.google-apps.document'
        # Force text/plain for MD/TXT to ensure conversion works
        source_mime = 'text/plain' 
        file_metadata['mimeType'] = target_mime
        
        # Remove extension from name for cleaner Doc title and set display name
        base_name = os.path.splitext(display_name)[0]
        file_metadata['name'] = base_name
    else:
        # Standard file
        file_metadata['name'] = display_name
    
    # Pass explicit source mime if needed (critical for conversion)
    media = MediaFileUpload(file_path, resumable=True, mimetype=source_mime)
    
    print(f"Uploading {display_name}...")
    try:
        # supportsAllDrives=True is critical if the target is a Shared Drive
        file = service.files().create(
            body=file_metadata,
            media_body=media,
            fields='id, webViewLink',
            supportsAllDrives=True
        ).execute()
        print(f"✅ Uploaded ({file.get('id')})")
        return file
    except Exception as e:
        print(f"❌ Failed to upload {name}: {e}")
        return None

def main():
    if len(sys.argv) < 2:
        print("Usage: python3 upload_to_drive.py <slug>")
        sys.exit(1)

    slug = sys.argv[1]
    today = datetime.date.today().strftime("%Y-%m-%d")
    folder_name = f"{today}_{slug}"
    
    # 1. Authenticate (User OAuth)
    service = authenticate()
    
    # 2. Create Subfolder
    if not PARENT_FOLDER_ID:
        print("Error: GOOGLE_DRIVE_FOLDER_ID not found in environment.")
        sys.exit(1)
        
    project_folder_id = create_folder(service, folder_name, PARENT_FOLDER_ID)
    
    # Standard Subfolders
    clips_folder_id = create_folder(service, "clips", project_folder_id)
    images_folder_id = create_folder(service, "images", project_folder_id)

    # Placeholder for clips
    placeholder_path = "DROP_VEO_CLIPS_HERE.txt"
    if not os.path.exists(placeholder_path):
        with open(placeholder_path, "w") as f:
            f.write("スマホからVeoで生成した動画（mp4）を、ファイル名の先頭を数字にしてここにアップロードしてください。\n例: 1.mp4, 2_bridge.mp4 ...")
    upload_file(service, placeholder_path, clips_folder_id)

    # 3. Upload Assets
    assets = [
        (f"public/videos/{slug}.mp4", project_folder_id),            # V3 Video
        (f"content/scripts/{slug}.md", project_folder_id),           # Script
        (f"content/prompts/{slug}_prompts.md", project_folder_id),   # Prompts
        (f"content/social/{slug}_posts.md", project_folder_id),      # Social Posts
        (f"content/articles/{slug}.md", project_folder_id)           # Article Draft
    ]
    
    # Sync Seed Images to images/ folder
    seed_pattern = f"projects/{slug}/images/*.png"
    seeds = glob.glob(seed_pattern)
    for seed in seeds:
        assets.append((seed, images_folder_id))

    # Sync Article WebP to images/ folder
    image_pattern = f"public/images/articles/{slug}/*.webp"
    articles_images = glob.glob(image_pattern)
    for img in articles_images:
        assets.append((img, images_folder_id))

    for asset_path, target_id in assets:
        if os.path.exists(asset_path):
            upload_file(service, asset_path, target_id)
        else:
            print(f"⚠️ Skipping missing asset: {asset_path}")

if __name__ == '__main__':
    try:
        main()
    except Exception as e:
        print(f"An error occurred: {e}")
        import traceback
        traceback.print_exc()

File: scripts/upload_to_drive.js
const fs = require('fs');
const path = require('path');
const { google } = require('googleapis');
require('dotenv').config({ path: '.env.local' });

// Configuration
const DRIVE_ROOT_ID = process.env.GOOGLE_DRIVE_FOLDER_ID;
const CREDENTIALS_PATH = path.resolve(__dirname, '../credentials.json');
const TOKEN_PATH = path.resolve(__dirname, '../token.json');

async function findOrCreateFolder(drive, parentId, folderName) {
    // Check if folder exists
    const query = `mimeType='application/vnd.google-apps.folder' and '${parentId}' in parents and name = '${folderName}' and trashed = false`;
    const res = await drive.files.list({
        q: query,
        fields: 'files(id, name)',
        spaces: 'drive',
    });

    if (res.data.files.length > 0) {
        console.log(`   📂 Found existing folder: ${folderName} (${res.data.files[0].id})`);
        return res.data.files[0].id;
    }

    // Create folder
    const fileMetadata = {
        name: folderName,
        mimeType: 'application/vnd.google-apps.folder',
        parents: [parentId],
    };
    const folder = await drive.files.create({
        requestBody: fileMetadata,
        fields: 'id',
    });
    console.log(`   zk Created new folder: ${folderName} (${folder.data.id})`);
    return folder.data.id;
}

async function uploadFile() {
    const filePath = process.argv[2];
    const subFolderName = process.argv[3]; // Optional: e.g. "Wealth Navigator Manifesto"
    const targetFileName = process.argv[4]; // Optional: e.g. "【ショート動画】Title.mp4"

    if (!filePath) {
        console.error("Usage: node scripts/upload_to_drive.js <file_path> [sub_folder_name] [target_file_name]");
        process.exit(1);
    }

    if (!fs.existsSync(filePath)) {
        console.error(`File not found: ${filePath}`);
        process.exit(1);
    }

    if (!DRIVE_ROOT_ID) {
        console.error("⚠️ GOOGLE_DRIVE_FOLDER_ID is missing in .env.local. Skipping upload.");
        process.exit(0);
    }

    if (!fs.existsSync(CREDENTIALS_PATH) || !fs.existsSync(TOKEN_PATH)) {
        console.error("❌ credentials.json or token.json not found in root.");
        console.error("   This script requires User OAuth credentials to bypass quota limits.");
        process.exit(1);
    }

    console.log(`🚀 Preparing upload...`);
    console.log(`   Source: ${filePath}`);

    try {
        // Auth Setup
        const content = fs.readFileSync(CREDENTIALS_PATH);
        const credentials = JSON.parse(content);
        const { client_secret, client_id, redirect_uris } = credentials.installed || credentials.web;

        const tokenContent = fs.readFileSync(TOKEN_PATH);
        const token = JSON.parse(tokenContent);

        const oAuth2Client = new google.auth.OAuth2(client_id, client_secret, redirect_uris ? redirect_uris[0] : 'http://localhost');
        oAuth2Client.setCredentials(token);

        const drive = google.drive({ version: 'v3', auth: oAuth2Client });

        // Determine destination folder
        // Determine destination folder
        // 1. Check for Tenant (DNA) Folder
        const dnaPath = path.resolve(__dirname, '../src/dna.config.json');
        let dnaName = null;
        try {
            const dna = require(dnaPath);
            dnaName = dna.identity?.name; // e.g., "Kominka Frontier"
        } catch (e) {
            // Ignore if DNA not found, use Root directly
        }

        let currentParentId = DRIVE_ROOT_ID;

        // 2. Switch to Tenant Folder if DNA is present
        if (dnaName) {
            console.log(`🧬 DNA Detected. Routing to Tenant Folder: [${dnaName}]`);
            currentParentId = await findOrCreateFolder(drive, DRIVE_ROOT_ID, dnaName);
        }

        // 3. Enter Subfolder if requested (e.g. "Shorts" or "Articles")
        if (subFolderName) {
            currentParentId = await findOrCreateFolder(drive, currentParentId, subFolderName);
        }

        const parentId = currentParentId;

        // Determine File Name
        const uploadName = targetFileName || path.basename(filePath);

        // Determine MIME Types for Conversion
        const ext = path.extname(filePath).toLowerCase();
        let mediaMimeType = 'application/octet-stream'; // Default
        let targetMimeType = null; // Default: same as media (no conversion)

        if (ext === '.mp4') mediaMimeType = 'video/mp4';
        else if (ext === '.png') mediaMimeType = 'image/png';
        else if (ext === '.jpg' || ext === '.jpeg') mediaMimeType = 'image/jpeg';
        else if (ext === '.txt') {
            mediaMimeType = 'text/plain';
            targetMimeType = 'application/vnd.google-apps.document'; // Convert to Doc
        } else if (ext === '.md') {
            mediaMimeType = 'text/markdown';
            targetMimeType = 'application/vnd.google-apps.document'; // Convert to Doc
        }

        const res = await drive.files.create({
            requestBody: {
                name: uploadName,
                parents: [parentId],
                mimeType: targetMimeType || undefined, // Set ONLY if converting
            },
            media: {
                mimeType: mediaMimeType,
                body: fs.createReadStream(filePath),
            },
            fields: 'id, name, webViewLink',
        });

        console.log(`✅ Upload Successful!`);
        console.log(`   Name: ${res.data.name}`);
        console.log(`   Folder ID: ${parentId}`);
        console.log(`   Link: ${res.data.webViewLink}`);

    } catch (error) {
        console.error("❌ Upload Failed:", error.message);
        if (error.message.includes('invalid_grant')) {
            console.log("⚠️ Token might be expired. Regenerate token.json.");
        }
    }
}

uploadFile();
